File: AdventOfCode/2023/day01/main.cpp
#include <bits/stdc++.h>
#include <cctype>
#include <string>

using namespace std;

int main() {
  // string FILENAME = "../inputs/1.txt";
  string FILENAME = "inputs/2.txt";
  ifstream file(FILENAME);
  string line;
  vector<string> inputs;
  while (getline(file, line)) {
    inputs.push_back(line);
  }

  // for (auto &c : inputs)
  //   cout << c << '\n';

  // PART 1 - Compute calibration value from first and last number concatenation
  int calibrationValue = 0;
  for (auto input : inputs) {
    vector<int> digits;
    // get all digits
    for (char l : input) {
      // cout << l;
      if (isdigit(l))
        digits.push_back(l - '0');
    }

    // get calibration value with first and last digits
    int val = digits[0] * 10 + digits.back();
    // cout << val << '\n';
    calibrationValue += val;
  }

  cout << "PART 1 - Calibration Value: " << calibrationValue << '\n';

  // PART 2 - Caliibration value spelled out with letter

  int calibrationValue2 = 0;
  map<string, int> wordNumberDict = {
      {"one", 1}, {"two", 2},   {"three", 3}, {"four", 4}, {"five", 5},
      {"six", 6}, {"seven", 7}, {"eight", 8}, {"nine", 9}, {"zero", 0},
  };

  for (auto t : inputs) {
    // get digits
    vector<int> digits;
    for (int i = 0; i < t.size(); i++) {
      if (isdigit(t[i])) {
        digits.push_back(t[i] - '0');
      }
      for (const auto &pair : wordNumberDict) {
        if ((t.size() >= i + pair.first.size()) &&
            (t.substr(i, pair.first.size()) == pair.first)) {
          digits.push_back(pair.second);
        }
      }
    }

    // compute calibration value
    int val = digits[0] * 10 + digits.back();
    calibrationValue2 += val;
  }

  cout << "PART 2 - Calibration Value: " << calibrationValue2 << '\n';

  return 0;
}

File: AdventOfCode/2023/day01/src/main.rs
use std::collections::HashMap;

fn main() {
    // 1. read inputs
    let inputs: Vec<&str> = include_str!("../inputs/2.txt")
        .lines()
        .collect();
    // println!("{:?}", inputs);

    // [ PART 1 - Compute calibrations values ]
    let calibrations: Vec<i32> = inputs.iter().map(|word| {
        // get first and last digit
        let first_digit = word.chars().find(|c| c.is_digit(10)).unwrap();
        let last_digit = word.chars().rev().find(|c| c.is_digit(10)).unwrap();

        // concat first and last digit
        let calibration_value = format!("{}{}", first_digit, last_digit)
            .parse::<i32>()
            .expect("Failed to parse calibration value");
        calibration_value
    }).collect();
    // println!("{:?}", calibrations);

    println!("PART 1 - Calibration Value: {}", calibrations.iter().sum::<i32>());


    // [ PART 2 - Compute calibrations values with maps]

    // create dictionary of mapping
    let mut dictionary = HashMap::new();
    dictionary.insert("one", "1");
    dictionary.insert("two", "2");
    dictionary.insert("three", "3");
    dictionary.insert("four", "4");
    dictionary.insert("five", "5");
    dictionary.insert("six", "6");
    dictionary.insert("seven", "7");
    dictionary.insert("eight", "8");
    dictionary.insert("nine", "9");
    dictionary.insert("zero", "0");

}

